pkg/hubble/metrics/api/api.go

```diff
@@ -64,56 +59,59 @@ type Handler interface {
 
 	// Status returns the configuration status of the metric handler
 	Status() string
-}
 
-// FlowProcessor is a metric handler which requires flows to perform metrics
-// accounting.
-// It is called upon receival of raw event data and is responsible
-// to perform metrics accounting according to the scope of the metrics plugin.
-type FlowProcessor interface {
+	// HandleConfigurationUpdate updates the metric handler configuration
+	HandleConfigurationUpdate(cfg *MetricConfig) error
+
+	// Deinit deregisters the metrics from the Prometheus registry
+	// and cleans up internal handler state
+	Deinit(registry *prometheus.Registry) error
+
 	// ProcessFlow must processes a flow event and perform metrics
 	// accounting
 	ProcessFlow(ctx context.Context, flow *pb.Flow) error
 }
 
-func NewHandlers(log logrus.FieldLogger, registry *prometheus.Registry, in []NamedHandler) (*Handlers, error) {
-	var handlers Handlers
-	for _, item := range in {
-		handlers.handlers = append(handlers.handlers, item.Handler)
-		if fp, ok := item.Handler.(FlowProcessor); ok {
-			handlers.flowProcessors = append(handlers.flowProcessors, fp)
-		}
-
-		if err := item.Handler.Init(registry, item.MetricConfig.ContextOptionConfigs); err != nil {
-			return nil, fmt.Errorf("unable to initialize metric '%s': %w", item.Name, err)
-		}
-
-		log.WithFields(logrus.Fields{
-			"name":   item.Name,
-			"status": item.Handler.Status(),
-		}).Info("Configured metrics plugin")
+func InitHandlers(log logrus.FieldLogger, registry *prometheus.Registry, in *[]NamedHandler) (*[]NamedHandler, error) {
+	var handlers []NamedHandler
+	for _, item := range *in {
+		InitHandler(log, registry, &item, &handlers)
 	}
 	return &handlers, nil
 }
 
-// ProcessFlow processes a flow by calling ProcessFlow it on to all enabled
+func InitHandler(log logrus.FieldLogger, registry *prometheus.Registry, item *NamedHandler, handlers *[]NamedHandler) error {
+	if err := item.Handler.Init(registry, item.MetricConfig); err != nil {
+		return fmt.Errorf("unable to initialize metric '%s': %w", item.Name, err)
+	}
+
+	*handlers = append(*handlers, *item)
+	log.WithFields(logrus.Fields{
+		"name":   item.Name,
+		"status": item.Handler.Status(),
+	}).Info("Configured metrics plugin")
+
+	return nil
+}
+
+// ExecuteAllProcessFlow processes a flow by calling ProcessFlow on all enabled
 // metric handlers
-func (h Handlers) ProcessFlow(ctx context.Context, flow *pb.Flow) error {
+func ExecuteAllProcessFlow(ctx context.Context, flow *pb.Flow, handlers *[]NamedHandler) error {
```

Comment:
This function is the same as the `(h Handlers) ProcessFlow` one it replaces, but instead of being implemented on a superfluous wrapper `type Handlers`, it just takes in the slice as an explicit argument (the underlying assembly is basically the same). It is not a new abstraction, rather, this PR removes a level of indirection of `type Handlers`, which IMO helps readability at several code sites.

I also don't feel that moving out this for simple loop would make it more easily understood; on the contrary, pervious feedback leaned towards making the Hubble start function less bloated, which this would go against.
If you think `ExecuteAllProcessFlow` doesn't express the intention clearly enough, perhaps renaming would achieve that?

Reply:
> I see no reason to pass a pointer to a slice here

Did you mean that a pass-by-value of a slice isn't too expensive, so I could be passing
`handlers []NamedHandler` instead of `handlers *[]NamedHandler`, since this code doesn't modify this argument?

Reply:
Ack: changed to use the for-loop in both static and dynamic code paths.
Moved this function out of the api for use as a test helper.